firedrake/interpolation.py

```diff
@@ -53,26 +58,92 @@ class Interpolator(object):
        arguments (such that they won't be collected until the
        :class:`Interpolator` is also collected).
     """
-    def __init__(self, expr, V, subset=None, access=op2.WRITE):
-        self.callable = make_interpolator(expr, V, subset, access)
+    def __init__(self, expr, V, subset=None, freeze_expr=False, access=op2.WRITE):
+        self.callable, args = make_interpolator(expr, V, subset, access)
+        self.args = args
+        self.nargs = len(args)
+        self.freeze_expr = freeze_expr
+        self.V = V
 
     @utils.known_pyop2_safe
-    def interpolate(self):
+    def interpolate(self, *function, output=None, transpose=False):
         """Compute the interpolation.
 
+        :arg function: If the expression being interpolated contains an
+           :class:`ufl.Argument`, then the :class:`.Function` value to
+           interpolate.
+        :kwarg output: Optional. A :class:`.Function` to contain the output.
+        :kwarg transpose: Set to true to apply the transpose (adjoint) of the
+           interpolation operator.
+
         :returns: The resulting interpolated :class:`.Function`.
         """
-        return self.callable()
+        if transpose and not self.nargs:
+            raise ValueError("Cannot currently apply transpose interpolation with no argument.")
+        if self.nargs != len(function):
+            raise ValueError("Passed %d Functions to interpolate, expected %d"
+                             % (len(function), self.nargs))
+        try:
+            callable = self.frozen_callable
```

Comment:
`callable` is sort of the wrong name for this thing, right?
That would suggest is a function that I can call that produces the result,
but it's actually the result.